core/src/test/java/org/infinispan/conflict/resolution/ConflictResolutionManagerTest.java

```diff
@@ -0,0 +1,265 @@
+package org.infinispan.conflict.resolution;
+
+import static org.infinispan.test.TestingUtil.wrapPerCacheInboundInvocationHandler;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotEquals;
+import static org.testng.Assert.fail;
+import static org.testng.AssertJUnit.assertTrue;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import org.infinispan.AdvancedCache;
+import org.infinispan.commands.remote.CacheRpcCommand;
+import org.infinispan.commands.remote.ClusteredGetCommand;
+import org.infinispan.commons.CacheException;
+import org.infinispan.configuration.cache.CacheMode;
+import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.container.entries.InternalCacheEntry;
+import org.infinispan.container.entries.InternalCacheValue;
+import org.infinispan.context.Flag;
+import org.infinispan.distribution.ch.ConsistentHash;
+import org.infinispan.partitionhandling.BasePartitionHandlingTest;
+import org.infinispan.remoting.inboundhandler.DeliverOrder;
+import org.infinispan.remoting.inboundhandler.PerCacheInboundInvocationHandler;
+import org.infinispan.remoting.inboundhandler.Reply;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.statetransfer.StateResponseCommand;
+import org.infinispan.test.TestingUtil;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+@Test(groups = "functional", testName = "conflict.resolution.ConflictResolutionManagerTest")
+public class ConflictResolutionManagerTest extends BasePartitionHandlingTest {
+
+   private static final String CACHE_NAME = "conflict-cache";
+   private static final int NUMBER_OF_OWNERS = 2;
+   private static final int NUMBER_OF_CACHE_ENTRIES = 100;
+   private static final int INCONSISTENT_VALUE_INCREMENT = 10;
+   private static final int NULL_VALUE_FREQUENCY = 20;
+
+   public ConflictResolutionManagerTest() {
+      this.cacheMode = CacheMode.DIST_SYNC;
+      this.partitionHandling = false;
+   }
+
+   @Override
+   protected void createCacheManagers() throws Throwable {
+      super.createCacheManagers();
+      ConfigurationBuilder builder = getDefaultClusteredCacheConfig(CacheMode.DIST_SYNC);
+      builder.clustering().partitionHandling().enabled(false).stateTransfer().fetchInMemoryState(true);
+      defineConfigurationOnAllManagers(CACHE_NAME, builder);
+   }
+
+   @Test(expectedExceptions = IllegalStateException.class,
+         expectedExceptionsMessageRegExp = ".* Unable to retrieve all versions of key .*")
+   public void testGetAllVersionsDuringStateTransfer() {
+      createAndSplitCluster();
+      blockStateTransferCompletion();
+      partition(0).merge(partition(1));
```

Comment:
Looking into `Partition.merge`, it seems that the code is waiting until the rebalance/state transfer completes. How can that complete, when you've blocked the state transfer? Can you be sure that the ST has already started?